### DISCUSSION TITLE
[Enhancement] [calendar] Calendar event overlays, single day, multiple day - see fullcalendarjs

### DISCUSSION THREAD

**Discussion by j-d-carmichael** [element-plus/element-plus#19347] [Feature Request]
Feature type: **`Enhancement`**

## Details

- Enhanced component name: `calendar`

## Description

https://fullcalendar.io/docs/theme-chooser-demo is the original lib from my perspective in this space and it would be amazing to see more of their features baked into the element-plus component lib. 

Just the basics from fullcal required to be able to use this element-plus cal in any app would be:
- add events to the cal. -> v-model
- display the events in the cal. from v-model
- listen when the user clicks an event on the calendar -> leave the dev to handle the click event 
- also distinguish when the click is just on the cal. and not the event -> leave the dev to handle the click event as is already 
- events added should be able to be a single day, or span over multiple days

In the docs of element-plus only the basics are there right now: https://element-plus.org/en-US/component/calendar.html#attributes

### Additional comments
Currently we cannot use the built in cal. as the basics for the cal. are missing.

**Comment by Dsaquel** [comment]
@j-d-carmichael Actually there is `update:modelValue` in the doc if you wanna listen on a value changed (implicitly gived by `v-model`)
There is also `@input` that doesnt documented but it is almost the same as `@update` does.
Maybe i dont understand the issue. Also can you explicitly tell what cal and fullcalendar mean please.
[playground](https://element-plus.run/#eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cCBsYW5nPVwidHNcIj5cbmltcG9ydCB7IHJlZiwgd2F0Y2hFZmZlY3QgfSBmcm9tICd2dWUnXG5cbmNvbnN0IGRhdGUgPSByZWY8RGF0ZT4oKVxuXG53YXRjaEVmZmVjdCgoKSA9PiB7XG4gIGlmKGRhdGUudmFsdWUpXG4gICAgY29uc29sZS5sb2coJ29yIHRyYWNrIGl0Jylcbn0pXG48L3NjcmlwdD5cblxuPHRlbXBsYXRlPlxuICA8ZWwtY2FsZW5kYXJcbiAgICB2LW1vZGVsPVwiZGF0ZVwiXG4gICAgQHVwZGF0ZTptb2RlbFZhbHVlPVwiKF9kYXRlKSA9PiBjb25zb2xlLmxvZygndXBkYXRlJylcIlxuICAgIEBpbnB1dD1cIihfZGF0ZSkgPT4gY29uc29sZS5sb2coJ2lucHV0JylcIlxuICAvPlxuPC90ZW1wbGF0ZT5cbiIsImVsZW1lbnQtcGx1cy5qcyI6ImltcG9ydCBFbGVtZW50UGx1cyBmcm9tICdlbGVtZW50LXBsdXMnXG5pbXBvcnQgeyBnZXRDdXJyZW50SW5zdGFuY2UgfSBmcm9tICd2dWUnXG5cbmxldCBpbnN0YWxsZWQgPSBmYWxzZVxuYXdhaXQgbG9hZFN0eWxlKClcblxuZXhwb3J0IGZ1bmN0aW9uIHNldHVwRWxlbWVudFBsdXMoKSB7XG4gIGlmIChpbnN0YWxsZWQpIHJldHVyblxuICBjb25zdCBpbnN0YW5jZSA9IGdldEN1cnJlbnRJbnN0YW5jZSgpXG4gIGluc3RhbmNlLmFwcENvbnRleHQuYXBwLnVzZShFbGVtZW50UGx1cylcbiAgaW5zdGFsbGVkID0gdHJ1ZVxufVxuXG5leHBvcnQgZnVuY3Rpb24gbG9hZFN0eWxlKCkge1xuICBjb25zdCBzdHlsZXMgPSBbJ2h0dHBzOi8vZmFzdGx5LmpzZGVsaXZyLm5ldC9ucG0vZWxlbWVudC1wbHVzQGxhdGVzdC9kaXN0L2luZGV4LmNzcycsICdodHRwczovL2Zhc3RseS5qc2RlbGl2ci5uZXQvbnBtL2VsZW1lbnQtcGx1c0BsYXRlc3QvdGhlbWUtY2hhbGsvZGFyay9jc3MtdmFycy5jc3MnXS5tYXAoKHN0eWxlKSA9PiB7XG4gICAgcmV0dXJuIG5ldyBQcm9taXNlKChyZXNvbHZlLCByZWplY3QpID0+IHtcbiAgICAgIGNvbnN0IGxpbmsgPSBkb2N1bWVudC5jcmVhdGVFbGVtZW50KCdsaW5rJylcbiAgICAgIGxpbmsucmVsID0gJ3N0eWxlc2hlZXQnXG4gICAgICBsaW5rLmhyZWYgPSBzdHlsZVxuICAgICAgbGluay5hZGRFdmVudExpc3RlbmVyKCdsb2FkJywgcmVzb2x2ZSlcbiAgICAgIGxpbmsuYWRkRXZlbnRMaXN0ZW5lcignZXJyb3InLCByZWplY3QpXG4gICAgICBkb2N1bWVudC5ib2R5LmFwcGVuZChsaW5rKVxuICAgIH0pXG4gIH0pXG4gIHJldHVybiBQcm9taXNlLmFsbFNldHRsZWQoc3R5bGVzKVxufVxuIiwidHNjb25maWcuanNvbiI6IntcbiAgXCJjb21waWxlck9wdGlvbnNcIjoge1xuICAgIFwidGFyZ2V0XCI6IFwiRVNOZXh0XCIsXG4gICAgXCJqc3hcIjogXCJwcmVzZXJ2ZVwiLFxuICAgIFwibW9kdWxlXCI6IFwiRVNOZXh0XCIsXG4gICAgXCJtb2R1bGVSZXNvbHV0aW9uXCI6IFwiQnVuZGxlclwiLFxuICAgIFwidHlwZXNcIjogW1wiZWxlbWVudC1wbHVzL2dsb2JhbC5kLnRzXCJdLFxuICAgIFwiYWxsb3dJbXBvcnRpbmdUc0V4dGVuc2lvbnNcIjogdHJ1ZSxcbiAgICBcImFsbG93SnNcIjogdHJ1ZSxcbiAgICBcImNoZWNrSnNcIjogdHJ1ZVxuICB9LFxuICBcInZ1ZUNvbXBpbGVyT3B0aW9uc1wiOiB7XG4gICAgXCJ0YXJnZXRcIjogMy4zXG4gIH1cbn1cbiIsIlBsYXlncm91bmRNYWluLnZ1ZSI6IjxzY3JpcHQgc2V0dXA+XG5pbXBvcnQgQXBwIGZyb20gJy4vQXBwLnZ1ZSdcbmltcG9ydCB7IHNldHVwRWxlbWVudFBsdXMgfSBmcm9tICcuL2VsZW1lbnQtcGx1cy5qcydcbnNldHVwRWxlbWVudFBsdXMoKVxuPC9zY3JpcHQ+XG5cbjx0ZW1wbGF0ZT5cbiAgPEFwcCAvPlxuPC90ZW1wbGF0ZT5cbiIsImltcG9ydC1tYXAuanNvbiI6IntcbiAgXCJpbXBvcnRzXCI6IHtcbiAgICBcInZ1ZVwiOiBcImh0dHBzOi8vZmFzdGx5LmpzZGVsaXZyLm5ldC9ucG0vQHZ1ZS9ydW50aW1lLWRvbUBsYXRlc3QvZGlzdC9ydW50aW1lLWRvbS5lc20tYnJvd3Nlci5qc1wiLFxuICAgIFwiQHZ1ZS9zaGFyZWRcIjogXCJodHRwczovL2Zhc3RseS5qc2RlbGl2ci5uZXQvbnBtL0B2dWUvc2hhcmVkQGxhdGVzdC9kaXN0L3NoYXJlZC5lc20tYnVuZGxlci5qc1wiLFxuICAgIFwiZWxlbWVudC1wbHVzXCI6IFwiaHR0cHM6Ly9mYXN0bHkuanNkZWxpdnIubmV0L25wbS9lbGVtZW50LXBsdXNAbGF0ZXN0L2Rpc3QvaW5kZXguZnVsbC5taW4ubWpzXCIsXG4gICAgXCJlbGVtZW50LXBsdXMvXCI6IFwiaHR0cHM6Ly9mYXN0bHkuanNkZWxpdnIubmV0L25wbS9lbGVtZW50LXBsdXNAbGF0ZXN0L1wiLFxuICAgIFwiQGVsZW1lbnQtcGx1cy9pY29ucy12dWVcIjogXCJodHRwczovL2Zhc3RseS5qc2RlbGl2ci5uZXQvbnBtL0BlbGVtZW50LXBsdXMvaWNvbnMtdnVlQDIvZGlzdC9pbmRleC5taW4uanNcIlxuICB9LFxuICBcInNjb3Blc1wiOiB7fVxufSIsIl9vIjp7fX0=)

**Reply by j-d-carmichael** [reply]
Hi @Dsaquel - apologies for the late reply

That clears up the click on the day - thanks.

Maybe It is not documented but, how can I add "events" to the calendar?

For example on a full calendar example you can see many events overlay onto the calendar view: https://fullcalendar.io/docs/theme-chooser-demo

But I cannot see how to do this with the element-plus component lib

**Reply by Dsaquel** [reply]
@j-d-carmichael i didnt know about fullcalendar lib really amazing lib right there, thanks for sharing.
At the moment `<el-calendar>` support only the event of "selecting a day" (`v-model`) and i think `<el-calendar>` is not ready to handle views, events, etc... as fullcalendar does right now.
But its definitely a really good inspiration for us for improving `<el-calendar>` and thanks again for that !

**Reply by j-d-carmichael** [reply]
Ah right ok, I was hoping for some hidden features there :)

Ok, we will continue to battle with fullcal in our element-plus app for now.

Thanks for the reply - I will modify this OP on this thread then

**Reply by Dsaquel** [reply]
@j-d-carmichael I dont know what you want for events but for maybe you can attach your wanted events in scoped-cell slot

[example](https://element-plus.run/#eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cCBsYW5nPVwidHNcIj5cbmltcG9ydCB7IHJlZiB9IGZyb20gJ3Z1ZSc7XG5cbmNvbnN0IGl0ZW1zID0gcmVmPFJlY29yZDxzdHJpbmcsIHN0cmluZz4+KHt9KSAvLyBkYXkgPT4gZGF0YVxuXG5sZXQgZGF0YUluY3IgPSAwXG5cbmZ1bmN0aW9uIG9uRHJhZ1N0YXJ0KGU6IERyYWdFdmVudCkge1xuICBlLmRhdGFUcmFuc2Zlci5zZXREYXRhKCdpdGVtJywgKGRhdGFJbmNyKyspLnRvU3RyaW5nKCkpXG4gIGUuZGF0YVRyYW5zZmVyLmVmZmVjdEFsbG93ZWQgPSAnbW92ZSdcbn1cblxuZnVuY3Rpb24gb25Ecm9wKGU6IERyYWdFdmVudCwgZGF5OiBzdHJpbmcpIHtcbiAgY29uc3QgZGF0YSA9IGUuZGF0YVRyYW5zZmVyLmdldERhdGEoJ2l0ZW0nKVxuICBpZighZGF0YSkgcmV0dXJuXG4gIGl0ZW1zLnZhbHVlW2RheV0gPSBkYXRhXG59XG48L3NjcmlwdD5cblxuPHRlbXBsYXRlPlxuICA8ZGl2IGRyYWdnYWJsZT1cInRydWVcIiBjbGFzcz1cIml0ZW1cIiBAZHJhZ3N0YXJ0PVwib25EcmFnU3RhcnRcIj5JIHdhbnQgdG8gYmUgYXNzaWduZWQgITwvZGl2PlxuICA8ZWwtY2FsZW5kYXI+XG4gICAgPHRlbXBsYXRlICNkYXRlLWNlbGw9XCJ7IGRhdGEgfVwiPlxuICAgICAgPGRpdiBcbiAgICAgICAgY2xhc3M9XCJkYXRhLWNlbGxcIiBcbiAgICAgICAgOmNsYXNzPVwiZGF0YS5pc1NlbGVjdGVkID8gJ2lzLXNlbGVjdGVkJyA6ICcnXCIgXG4gICAgICAgIEBkcmFnb3Zlcj1cIihlKSA9PiBlLnByZXZlbnREZWZhdWx0KClcIiBcbiAgICAgICAgQGRyb3A9XCIoZSkgPT4gb25Ecm9wKGUsIGRhdGEuZGF5KVwiXG4gICAgICA+XG4gICAgICAgIHt7IGRhdGEuZGF5LnNwbGl0KCctJykuc2xpY2UoMSkuam9pbignLScpIH19XG4gICAgICAgIHt7IGRhdGEuaXNTZWxlY3RlZCA/ICfinJTvuI8nIDogJycgfX1cbiAgICAgICAgPGRpdiB2LWlmPVwiaXRlbXNbZGF0YS5kYXldXCIgY2xhc3M9XCJpdGVtXCIgQGNsaWNrLnN0b3A9XCJkZWxldGUgaXRlbXNbZGF0YS5kYXldXCI+XG4gICAgICAgICAge3sgaXRlbXNbZGF0YS5kYXldIH19XG4gICAgICAgICAgLSBcbiAgICAgICAgICBjbGljayB0byBjbGVhclxuICAgICAgICA8L2Rpdj5cbiAgICAgIDwvZGl2PlxuICAgIDwvdGVtcGxhdGU+XG4gIDwvZWwtY2FsZW5kYXI+XG48L3RlbXBsYXRlPlxuXG48c3R5bGU+XG4uZGF0YS1jZWxsIHtcbiAgd2lkdGg6IDEwMCU7XG4gIGhlaWdodDogMTAwJTtcbn1cblxuLmlzLXNlbGVjdGVkIHtcbiAgY29sb3I6ICMxOTg5ZmE7XG59XG5cbi5pdGVtIHtcbiAgZGlzcGxheTogZmxleDtcbiAgYmFja2dyb3VuZDogcmVkO1xufVxuPC9zdHlsZT5cbiIsImVsZW1lbnQtcGx1cy5qcyI6ImltcG9ydCBFbGVtZW50UGx1cyBmcm9tICdlbGVtZW50LXBsdXMnXG5pbXBvcnQgeyBnZXRDdXJyZW50SW5zdGFuY2UgfSBmcm9tICd2dWUnXG5cbmxldCBpbnN0YWxsZWQgPSBmYWxzZVxuYXdhaXQgbG9hZFN0eWxlKClcblxuZXhwb3J0IGZ1bmN0aW9uIHNldHVwRWxlbWVudFBsdXMoKSB7XG4gIGlmIChpbnN0YWxsZWQpIHJldHVyblxuICBjb25zdCBpbnN0YW5jZSA9IGdldEN1cnJlbnRJbnN0YW5jZSgpXG4gIGluc3RhbmNlLmFwcENvbnRleHQuYXBwLnVzZShFbGVtZW50UGx1cylcbiAgaW5zdGFsbGVkID0gdHJ1ZVxufVxuXG5leHBvcnQgZnVuY3Rpb24gbG9hZFN0eWxlKCkge1xuICBjb25zdCBzdHlsZXMgPSBbJ2h0dHBzOi8vY2RuLmpzZGVsaXZyLm5ldC9ucG0vZWxlbWVudC1wbHVzQGxhdGVzdC9kaXN0L2luZGV4LmNzcycsICdodHRwczovL2Nkbi5qc2RlbGl2ci5uZXQvbnBtL2VsZW1lbnQtcGx1c0BsYXRlc3QvdGhlbWUtY2hhbGsvZGFyay9jc3MtdmFycy5jc3MnXS5tYXAoKHN0eWxlKSA9PiB7XG4gICAgcmV0dXJuIG5ldyBQcm9taXNlKChyZXNvbHZlLCByZWplY3QpID0+IHtcbiAgICAgIGNvbnN0IGxpbmsgPSBkb2N1bWVudC5jcmVhdGVFbGVtZW50KCdsaW5rJylcbiAgICAgIGxpbmsucmVsID0gJ3N0eWxlc2hlZXQnXG4gICAgICBsaW5rLmhyZWYgPSBzdHlsZVxuICAgICAgbGluay5hZGRFdmVudExpc3RlbmVyKCdsb2FkJywgcmVzb2x2ZSlcbiAgICAgIGxpbmsuYWRkRXZlbnRMaXN0ZW5lcignZXJyb3InLCByZWplY3QpXG4gICAgICBkb2N1bWVudC5ib2R5LmFwcGVuZChsaW5rKVxuICAgIH0pXG4gIH0pXG4gIHJldHVybiBQcm9taXNlLmFsbFNldHRsZWQoc3R5bGVzKVxufVxuIiwidHNjb25maWcuanNvbiI6IntcbiAgXCJjb21waWxlck9wdGlvbnNcIjoge1xuICAgIFwidGFyZ2V0XCI6IFwiRVNOZXh0XCIsXG4gICAgXCJqc3hcIjogXCJwcmVzZXJ2ZVwiLFxuICAgIFwibW9kdWxlXCI6IFwiRVNOZXh0XCIsXG4gICAgXCJtb2R1bGVSZXNvbHV0aW9uXCI6IFwiQnVuZGxlclwiLFxuICAgIFwidHlwZXNcIjogW1wiZWxlbWVudC1wbHVzL2dsb2JhbC5kLnRzXCJdLFxuICAgIFwiYWxsb3dJbXBvcnRpbmdUc0V4dGVuc2lvbnNcIjogdHJ1ZSxcbiAgICBcImFsbG93SnNcIjogdHJ1ZSxcbiAgICBcImNoZWNrSnNcIjogdHJ1ZVxuICB9LFxuICBcInZ1ZUNvbXBpbGVyT3B0aW9uc1wiOiB7XG4gICAgXCJ0YXJnZXRcIjogMy4zXG4gIH1cbn1cbiIsIlBsYXlncm91bmRNYWluLnZ1ZSI6IjxzY3JpcHQgc2V0dXA+XG5pbXBvcnQgQXBwIGZyb20gJy4vQXBwLnZ1ZSdcbmltcG9ydCB7IHNldHVwRWxlbWVudFBsdXMgfSBmcm9tICcuL2VsZW1lbnQtcGx1cy5qcydcbnNldHVwRWxlbWVudFBsdXMoKVxuPC9zY3JpcHQ+XG5cbjx0ZW1wbGF0ZT5cbiAgPEFwcCAvPlxuPC90ZW1wbGF0ZT5cbiIsImltcG9ydC1tYXAuanNvbiI6IntcbiAgXCJpbXBvcnRzXCI6IHtcbiAgICBcInZ1ZVwiOiBcImh0dHBzOi8vY2RuLmpzZGVsaXZyLm5ldC9ucG0vQHZ1ZS9ydW50aW1lLWRvbUBsYXRlc3QvZGlzdC9ydW50aW1lLWRvbS5lc20tYnJvd3Nlci5qc1wiLFxuICAgIFwiQHZ1ZS9zaGFyZWRcIjogXCJodHRwczovL2Nkbi5qc2RlbGl2ci5uZXQvbnBtL0B2dWUvc2hhcmVkQGxhdGVzdC9kaXN0L3NoYXJlZC5lc20tYnVuZGxlci5qc1wiLFxuICAgIFwiZWxlbWVudC1wbHVzXCI6IFwiaHR0cHM6Ly9jZG4uanNkZWxpdnIubmV0L25wbS9lbGVtZW50LXBsdXNAbGF0ZXN0L2Rpc3QvaW5kZXguZnVsbC5taW4ubWpzXCIsXG4gICAgXCJlbGVtZW50LXBsdXMvXCI6IFwiaHR0cHM6Ly9jZG4uanNkZWxpdnIubmV0L25wbS9lbGVtZW50LXBsdXNAbGF0ZXN0L1wiLFxuICAgIFwiQGVsZW1lbnQtcGx1cy9pY29ucy12dWVcIjogXCJodHRwczovL2Nkbi5qc2RlbGl2ci5uZXQvbnBtL0BlbGVtZW50LXBsdXMvaWNvbnMtdnVlQDIvZGlzdC9pbmRleC5taW4uanNcIlxuICB9LFxuICBcInNjb3Blc1wiOiB7fVxufSIsIl9vIjp7fX0=) for item to drag and drop. You can find work around like that i think. Give me a feedback if you dont see any workaround for your event

**Reply by j-d-carmichael** [reply]
That's pretty nice, but we just need to build a calendar like google cal. 

Pull a series of events from an api, load them onto the calendar UI.

The event objects from the api would have a start and end date, js date object.

The cal. would then add these events automatically, basically the same as fullcaledar (as seem in google cal outlook).

We will stick with fullcalendar for now.

Thanks.
John
